### PR TITLE
Add reduce-scatter coalescing for FSDP/ZeRO1

### PR DESCRIPTION
**Pull request by jeffhataws**
This PR adds reduce-scatter coalescence support and use that in FSDP/ZeRO1 (replacing https://github.com/pytorch/xla/pull/5938). This also enables using reduce-scatter's scale param in FSD.. This PR is companion to https://github.com/pytorch/xla/pull/5950 and to be used in conjunction with openxla https://github.com/openxla/xla/pull/5740 .

This is a revival of https://github.com/pytorch/xla/pull/4145 . Will need to address the comments.

### CODE REVIEW

File: torch_xla/csrc/cross_replica_reduces.cpp
Comment (JackCaoG) [inline torch_xla/csrc/cross_replica_reduces.cpp:278]: can we do the similar thing here? not changing the default behavior of the reducescatter?

File: torch_xla/core/xla_model.py
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:772]: diito, use `elif` here.

**Comment by JackCaoG** [comment]
I think my comment for this pr will be very similar to the all-gather one, let's try not to change the default behavior of `reduce_scatter`. Let me know when I should do another round of review

File: torch_xla/csrc/cross_replica_reduces.cpp
Comment (jeffhataws) [reply torch_xla/csrc/cross_replica_reduces.cpp:313]: fixed

File: torch_xla/core/xla_model.py
Comment (jeffhataws) [reply torch_xla/core/xla_model.py]: fixed

**Comment by JackCaoG** [comment]
Now that allgather one is merged, do you mind resolve the conflict in this pr?

**Comment by jeffhataws** [comment]
FSDP test passes with pin layout disabled (--no_pin_layout_in_collective_ops).

@JackCaoG  any idea on where this error comes from or how to resolve it?

`F0000 00:00:1701586053.119077  375805 shape.h:169] Check failed: has_layout() element_type: TUPLE tuple_shapes { element_type: S64 dimensions: 4 layout { minor_to_major: 0 } is_dynamic_dimension: false } tuple_shapes { element_type: S64 dimensions: 10 layout { minor_to_major: 0 } is_dynamic_dimension: false }  `

**Comment by jeffhataws** [comment]
> FSDP test passes with pin layout disabled (--no_pin_layout_in_collective_ops).
> 
> @JackCaoG any idea on where this error comes from or how to resolve it?
> 
> `F0000 00:00:1701586053.119077 375805 shape.h:169] Check failed: has_layout() element_type: TUPLE tuple_shapes { element_type: S64 dimensions: 4 layout { minor_to_major: 0 } is_dynamic_dimension: false } tuple_shapes { element_type: S64 dimensions: 10 layout { minor_to_major: 0 } is_dynamic_dimension: false } `

I worked around this by falling back to single-tensor reduce-scatter when not coalescing (bucket size is 0). Now the FSDP test is passing on GPU.

**Comment by JackCaoG** [comment]
somewhere in the code that xla is assert on input shape has layout, but tuple does not have layout... This is the part where I say we should actually test coalescing on fsdp resnet test. If there are errors on XLA side we should fix them. If it doesn't work on GPU and TPU, for this release we can call it a Trainium specified feature and trying to fix them for next release.

**Comment by jeffhataws** [comment]
> somewhere in the code that xla is assert on input shape has layout, but tuple does not have layout... This is the part where I say we should actually test coalescing on fsdp resnet test. If there are errors on XLA side we should fix them. If it doesn't work on GPU and TPU, for this release we can call it a Trainium specified feature and trying to fix them for next release.

Yes I agree. Will you merge this for 2.2 then?

**Comment by JackCaoG** [comment]
I can merge to unblock you guys(and since it doesn't impact the default behavior), but if we can't verify that it actually works I am just not going to mention it in the release note.

File: torch_xla/core/xla_model.py
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:794]: wouldn't this error make `output != None` logic never being run?

File: torch_xla/csrc/cross_replica_reduces.cpp
Comment (JackCaoG) [inline torch_xla/csrc/cross_replica_reduces.cpp:395]: I think this part of the logic is correct.. but it is really hard for me to be sure without a proper test case in https://github.com/pytorch/xla/blob/master/test/test_mp_reduce_scatter.py. Let me try to enable that test on GPU first...

File: torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py
Comment (JackCaoG) [inline torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py:409]: this prs changes default behavior of the FSDP, if you want to merge this part, please make sure proper tests are added. @alanwaketan can you review the FSDP part of codes?

**Comment by JackCaoG** [comment]
@jeffhataws can you separate out the FSDP change in a separate pr? I can asked @alanwaketan to review that part. I will try to help you to land the reduce scatter one after adding some tests.

File: torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py
Comment (jeffhataws) [reply torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py]: Will post a separate PR for the FSDP change.

File: torch_xla/core/xla_model.py
Comment (jeffhataws) [reply torch_xla/core/xla_model.py]: Fixed. Thanks for the catch.

**Comment by jeffhataws** [comment]
Strange, locally ./test/run_tests.sh is failing after latest changes to separate FSDP change out:

```
======================================================================
FAIL: test_patched_linear_2D_bias (__main__.TestAtenXlaTensor)
----------------------------------------------------------------------
Traceback (most recent call last):                                                                
  File "/home/ubuntu/jthuynh/pytorch/xla/test/test_operations.py", line 1817, in test_patched_linear_2D_bias                                                                                         
    self.assertTrue(torch.allclose(output.cpu(), output_cpu))                                     
AssertionError: False is not true     
                                                                                                  
======================================================================
FAIL: test_patched_linear_3D (__main__.TestAtenXlaTensor)
----------------------------------------------------------------------
Traceback (most recent call last):                                                                
  File "/home/ubuntu/jthuynh/pytorch/xla/test/test_operations.py", line 1761, in test_patched_linear_3D                                                                                              
    self.assertTrue(torch.allclose(output.cpu(), output_cpu))                                     
AssertionError: False is not true         
                                                                                                  
======================================================================
FAIL: test_patched_linear_3D_bias (__main__.TestAtenXlaTensor)                                    
----------------------------------------------------------------------                            
Traceback (most recent call last):      
  File "/home/ubuntu/jthuynh/pytorch/xla/test/test_operations.py", line 1789, in test_patched_linear_3D_bias                                                                                         
    self.assertTrue(torch.allclose(output.cpu(), output_cpu))
AssertionError: False is not true                                                                                                                                                                    
                                        
```

**Comment by JackCaoG** [comment]
hmm can you print out output values? If they are close you can ignore them. Might just be a precision issue and your local machine and CI are using different random seeds.

**Comment by jeffhataws** [comment]
> hmm can you print out output values? If they are close you can ignore them. Might just be a precision issue and your local machine and CI are using different random seeds.

Something weird with my environtment so let's ignore. CI seems to be fine.

**Comment by jeffhataws** [comment]
> @jeffhataws can you separate out the FSDP change in a separate pr? I can asked @alanwaketan to review that part. I will try to help you to land the reduce scatter one after adding some tests.

https://github.com/pytorch/xla/pull/6024 is the PR for FSDP.

File: torch_xla/core/xla_model.py
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:801]: this part seems wrong, if we are in here, it means that input is not a tensor nor a list. Shouldn't this check be in side the 
```
elif isinstance(input, list) and all(
      isinstance(v, torch.Tensor) for v in input):
```
and handle the `out != None` for the list case?

Comment (JackCaoG) [reply torch_xla/core/xla_model.py:801]: I am a bit conufse here. I guess you want to handle the `output != None` case for `reduce_scatter` and list input since C++ side there is implementation for that, but in here it is a no-op(even if you move this logic under the `elif isinstance(input, list)`)

File: torch_xla/csrc/init_python_bindings.cpp
Comment (JackCaoG) [inline torch_xla/csrc/init_python_bindings.cpp:248]: if you want to handle the `outputs` cases, it is better to define a `ReduceScatterCoalescedOut` instead of merging the logic into a single function.

**Comment by JackCaoG** [comment]
Left some comments, @jeffhataws can you rebase this pr and add test cases to https://github.com/pytorch/xla/blob/master/test/test_mp_reduce_scatter.py so we can actually run `reduce_scatter` with list input on GPU?

File: torch_xla/core/xla_model.py
Comment (jeffhataws) [reply torch_xla/core/xla_model.py:795]: Thanks! Fixed.

**Comment by jeffhataws** [comment]
> Left some comments, @jeffhataws can you rebase this pr and add test cases to https://github.com/pytorch/xla/blob/master/test/test_mp_reduce_scatter.py so we can actually run `reduce_scatter` with list input on GPU?

Added a test case. Only can handle pin_layout=False at the moment.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jeffhataws) [reply torch_xla/csrc/init_python_bindings.cpp:248]: Is it ok if I work on this in another change?

**Comment by JackCaoG** [comment]
Let me take another look. 2.2 branch is cut, I will take care of bakporting this pr after it is merged.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (JackCaoG) [reply torch_xla/csrc/init_python_bindings.cpp:248]: we can work on the `reduce_scatter_out` in a separate pr

**Review by JackCaoG — COMMENTED** [review]
LGTM with one minor comment. If we can explictly error out on `out != None` case and CI pass I can merge this one.

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jeffhataws) [reply torch_xla/csrc/init_python_bindings.cpp:248]: Thanks Jack. Here's the change to error out if output!=None: https://github.com/pytorch/xla/pull/5956/commits/9baadf50994a77cc578f5166304484a4c98ab0ac

**Review by JackCaoG — APPROVED** [review]
Thanks @jeffhataws !

File: torch_xla/csrc/init_python_bindings.cpp
Comment (jeffhataws) [reply torch_xla/csrc/init_python_bindings.cpp:248]: https://github.com/pytorch/xla/pull/6058 is for adding out-of-place version for reduce-scatter, along with https://github.com/pytorch/xla/pull/6059 for out-of-place all-gather.

**Pull request merged**
